Test framework student test run

jeremygoh opened the pull request:
Work in progress. This is a fairly big and verbose PR - it's mostly for visibility @david-yz-liu @reidka  into what I've been working on recently. As you'll see from the below description, there is a significant amount to be fixed. On the bright side, I was able to get a test to run and save + display its results (after a page refresh - the js is broken).


Work done
- re-worked a lot of the test helper code. 
- lots of debugging. Got rid of variables that did nothing, non-existent method calls that caused runtime errors etc
- now able to execute a test script and save it


Outstanding issues
- there is a terrible hack on line `app/helpers/automated_tests_helper.rb` to get past the validation requiring a submission association for a test script result. There are lots of references and uses of submissions in the code it is never passed in. If this is an association we need (which I think it is), then this will need to be fixed - either in this PR or the next one

- can't merge this automatically. I checked out test-framework branch and it currently doesn't work for file uploading, there's an error with calling testfiles.each ... where testfiles is nil. @WilsonChiang - is this a known issue. Once this is addressed, I'll merge test-framework here so that this PR can be automatically merged

- names of test scripts uploaded are not being saved properly, should be of the form "hello.rb"

- need to verify that options for test scripts such as "run on request" are being saved correctly (this can be looked into at the same time as the test scripts upload issue)

- javascript for the display of test results after a student runs them is missing. I haven't yet dug into the extent of this problem.

![test_results](https://cloud.githubusercontent.com/assets/4348811/11056622/a569534c-8738-11e5-825e-1c739135933f.png)
![livepipe js is missing](https://cloud.githubusercontent.com/assets/4348811/11056620/a552ff84-8738-11e5-966a-32a4ac4365cf.jpg)
![test script upload error](https://cloud.githubusercontent.com/assets/4348811/11056621/a55612fa-8738-11e5-964d-f6ae546658e9.jpg)

jeremygoh commented:
Hound is complaining about the style of the code that was previously written (and which I've commented out or sourced from previous check ins). In addition, some of the mechanisms of the code previously written for this feature need to be revisited seriously (e.g. raising runtime errors as a method of passing failure and *success* messages to the client). 

My goal is to get as much working code checked in for this feature and try and circle back to refactor. Where I write additional code, I'll endeavour to maintain the style standards for this project.

jeremygoh commented:
Ahh, my mistake on the test file upload error. I didn't realize a change hadn't yet been merged into the branch! Test file upload seems ok now.

david-yz-liu commented on app/controllers/automated_tests_controller.rb:
```diff
@@ -36,15 +36,15 @@ def update
-      #   render :manage
-      #   return
-      # end
+      begin
```
Delete this block. There is already a call to process_test_form above. If there's an error, let it blow up for now.

david-yz-liu commented:
@jeremygoh I need very clear instructions on how to "run a test file" and how to see the results.

david-yz-liu commented on app/helpers/automated_tests_helper.rb:
```diff
@@ -480,6 +486,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+    # sort list_run_scripts using ruby's in place sorting method
+    list_run_scripts.sort_by! {|script| script.seq_num}
+
+    # list_run_scripts should be sorted now. Perform a check here.
```
Remove this block of code.

david-yz-liu commented on app/helpers/automated_tests_helper.rb:
```diff
@@ -480,6 +486,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+    @group = @grouping.group
+    @repo_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, @group.repo_name)
+
+    @list_of_servers = MarkusConfigurator.markus_ate_test_server_hosts.split(' ')
```
Yeah, I don't want to see any mention of external servers at this point. What I really want is to see the tests run successfully on the server.

david-yz-liu commented on app/helpers/automated_tests_helper.rb:
```diff
@@ -480,6 +486,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+  # using round-robin. Return the id of the server, and return -1
+  # if no server is available.
+  # TODO: keep track of the max num of tests running on a server
+  def self.choose_test_server()
```
Get rid of this.

david-yz-liu commented on app/helpers/automated_tests_helper.rb:
```diff
@@ -480,6 +486,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+  # stdout or stderr, depending on whether the execution passed or
+  # had error; the second one is a boolean variable, true => execution
+  # passeed, false => error occurred.
+  def self.launch_test(server_id, assignment, repo_dir, call_on)
```
You have my blessing to rewrite this method. Do three things:

1. Copy all the necessary files into a folder. You can pick the folder name.
2. Run the tests (I don't care if it's just a single test file) and redirect stdout and stderr to two different files.
3. Return the contents of the files, separately.

david-yz-liu commented on app/helpers/automated_tests_helper.rb:
```diff
@@ -480,6 +486,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+  end
+
+  def self.process_result(result, grouping_id, assignment_id)
+    parser = XML::Parser.string(result)
```
I don't want to see this method yet. Remove it from this pull request. If you can save the contents of the files into the database (no parsing), I'll be happy.

david-yz-liu commented on app/views/layouts/_sub_menu.html.erb:
```diff
@@ -234,6 +234,12 @@
         </li>
+        <li class='<%= "active" if controller.controller_name == 'automated_tests' %>'>
+          <%= link_to I18n.t("automated_tests.title"),
+                      :controller => 'automated_tests',
```
Use the new hash syntax `controller: 'automated_tests`

david-yz-liu commented:
There is a new file `dump.rdb` which I assume should not be included.

jeremygoh commented:
@david-yz-liu , I've made the changes you recommended. 

- deleted unneeded blocks
- switch to hash "key: value" syntax
- commented out code involving server selection, just using 'localhost' for now
- refactored `launch_test` method, it wasn't too substantial of a refactor. The main thing is it now returns stderr, stdout, and status rather than just result and status. It also now writes stderr and stdout if the test was successfully run to error.txt and output.txt files located in a folder suffixed with unix time `#{AUTOMATED_TESTS_REPOSITORY}/test_runs/test_run_#{Time.now.to_i}`
- stubbed `process_result` method
- removed dump.rdb (redis database of jobs), and added it to the .gitignore file

Notes on running tests
- test script uploading is not working. It creates test scripts with an incorrect value for `script_name`, it is currently saving an ActiveRecord object as the file name (a serialized version I think) rather than something like `rb_suite_1.rb`.
- test scripts are also not being saved into the correct repo directory.

To see a sample automated test run:
- make a submission as a student, e.g. `Hello.rb` from the sample automated test files
- upload test script as a teacher, e.g. `rb_suite_1.rb` from the sample automated test files
- Modify the last TestScript record's name to reflect the file name of the test script just uploaded. 
```
x = TestScript.last
x.script_name = 'rb_suite_1.rb'
x.save
```
- copy the above file uploaded into the correct repo, e.g. 
    data/dev/automated_tests/A3/rb_suite_1.rb
- run `redis-server`
- run `VVERBOSE=1 QUEUE=test_waiting_list rake environment resque:work`
- run `rails s`

david-yz-liu commented:
@jeremygoh Okay, so I got everything up and running, I logged in as the student and hit "Run Tests". I see some output from resque, which seems like things worked. However, I don't see anything in the browser; is the test output shown somewhere else?

jeremygoh commented:
The test output is not currently returned to the browser. I believe it's
because the old code used something called livepipe.js as part of the Ajax
call, and this library is not currently present in the code.

I'll dig into that.
On 18 Nov 2015 7:12 pm, "David Liu" <notifications@github.com> wrote:

> @jeremygoh <https://github.com/jeremygoh> Okay, so I got everything up
> and running, I logged in as the student and hit "Run Tests". I see some
> output from resque, which seems like things worked. However, I don't see
> anything in the browser; is the test output shown somewhere else?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MarkUsProject/Markus/pull/2348#issuecomment-157935409>
> .
>

david-yz-liu commented:
That's fine, I was just curious. Where can I see the test results then?

jeremygoh commented:
@david-yz-liu  you can see the test results at
`#{AUTOMATED_TESTS_REPOSITORY}/test_runs/test_run_#{Time.now.to_i}`

which on dev would look something like
`data/dev/automated_tests/test_runs/test_run_1447820641`
With the numerical suffix depending on the unix time in seconds.

In that folder there should be an output.txt and an error.txt. The results of the test are not currently being returned to the browser but I can look at that next?

Also here's an example output file I generated with a test_run
[output.txt](https://github.com/MarkUsProject/Markus/files/39199/output.txt)

david-yz-liu commented:
@jeremygoh I was wondering about the line lengths... I wonder if Hound just gives up after a certain number of errors.

I haven't checked the output file myself yet, but it sounds plausible. You can work on displaying the output to the student next.

jeremygoh commented:
Ok. Will do!

I'll keep an eye on the line lengths as well.
On 19 Nov 2015 08:55, "David Liu" <notifications@github.com> wrote:

> Merged #2348 <https://github.com/MarkUsProject/Markus/pull/2348>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MarkUsProject/Markus/pull/2348#event-469395371>.
>
